DM-30224: Refactor GAaP plugin

arunkannawadi opened the pull request:
The main rationale for refactoring the base class for the GAaP as a mixin as opposed to a GenericPlugin is to custom wrap the base class and define specialized `measure` methods for DM-29290 that is not possible by inheriting from GenericPlugin. The last commit captures most of the refactoring, while other commits are related to cleanups in getting the plugins to near completion.

MorganSchmitz commented on python/lsst/meas/extensions/gaap/_gaap.py:
```diff
@@ -388,89 +404,293 @@ def _convolve(self, exposure: afwImage.Exposure, modelPsf: afwDetection.Gaussian
+                    continue
+
+                aperSigma2 = sigma**2 - targetSigma**2
+                aperShape = afwGeom.Quadrupole(aperSigma2, aperSigma2, 0.0)
```
As far as I can tell, before the change we had `normalize=True` passed on to `Quadrupole` here? Why this change?

arunkannawadi replied:
The `normalize=True` option is no longer required since we get here only if `aperSigma2` is positive. It wasn't necessarily the case previously (due to complex control flow) so I had implement the check here. Btw, `normalize=True` checks if the `Quadrupole` object has valid parameters and throws an `InvalidParameterError` if not.

MorganSchmitz commented on python/lsst/meas/extensions/gaap/_gaap.py:
```diff
@@ -388,89 +404,293 @@ def _convolve(self, exposure: afwImage.Exposure, modelPsf: afwDetection.Gaussian
+        baseName : `str`
+            The base name of the GAaP field for which the flag must be set.
+        """
+        flagKey = measRecord.schema.join(baseName, "flag_bigpsf")
```
Ah, sorry, I was going through commits one after the other. This actually makes perfect sense.

I would still suggest changing the name of the function to something less general than `_setFlag`!

arunkannawadi replied:
When I came up with flag definitions, I wasn't sure if I had exhausted all possible flags. Now I'm fairly certain I wouldn't need more flags and I can make this just `flag` instead of `flag_bigpsf`.

TallJimbo replied:
It's always good to leave some room for future flags without having to change the meaning of old ones.

arunkannawadi replied:
I've kept the flag name and the function name the same, but allow `flagName` to be passed (which is "bigpsf" by default now). So the generic function name is justified while the use case is specific to one flag.

arunkannawadi replied:
Coming back on flag suffixes, during the aperture correction stage, it expects the flag column to end with `_flag`. Currently, this is the only flag column for each `instFlux`. I can either rename this to `_flag` instead of `_flag_bigpsf` and set any other failure modes in the future to also set the `_flag`, or have a separate `_flag` field that is set if any of the flags (currently only `_flag_bigpsf`) is set. Thoughts?

TallJimbo replied:
Preferred pattern is definitely the latter: have both a generic flag and specific flags, even if there is only one of the later to start with.

arunkannawadi replied:
Thanks Jim. I was worried about proliferation of flag fields, but we're only talking about increasing it from 3 to 6 with the default settings, which is not bad. I'll incorporate that in the aperture correction branch then.

MorganSchmitz reviewed (commented):
A few minor comments on the small commits; I still have to go through the mixin one, but looks good so far!

arunkannawadi commented on python/lsst/meas/extensions/gaap/_gaap.py:
```diff
@@ -78,46 +81,54 @@ def _isOdd(x: int) -> bool:  # noqa: N805
         dtype=float,
         default=[4.0, 5.0],
-        doc="List of sigmas (in pixels) of circular Gaussian apertures to apply on "
+        doc=("List of sigmas (in pixels) of circular Gaussian apertures to apply on "
```
Multi-line docs for config fields is rare, and I have seen examples with and without parenthesis in the codebase. I couldn't find an explicit instruction in the dev-guide and I felt safer putting them within parenthesis. Happy to remove it if that's preferred.

arunkannawadi replied:
I also don't see what extra whitespace you're referring to. Can you suggest the change you're referring to directly on the PR?

timj replied:
As far as I can tell, since this is already inside parentheses the opening bracket is not needed for that string. A quick test of:

```python
def a(b):
    print(b)

a(b="a long string"
    " that continues")
```
does work fine without needing the extra `()`

arunkannawadi replied:
Right, I was just uneasy seeing strings concatenated just like that. I did have the multi-line docs without extra parenthesis but given the general rule in the dev-guide that we should use minimal parenthesis required, I'll go back to not having those extra parenthesis.

TallJimbo commented on python/lsst/meas/extensions/gaap/_gaap.py:
```diff
@@ -356,6 +359,11 @@ def _convolve(self, exposure: afwImage.Exposure, modelPsf: afwDetection.Gaussian
+        Notes
+        -----
+        During normal mode of operation, ``modelPsf`` is intended to be of the
+        type`~lsst.afw.detection.GaussianPsf`, this is not enforced.
```
```suggestion
        type `~lsst.afw.detection.GaussianPsf`, this is not enforced.
```

TallJimbo commented on python/lsst/meas/extensions/gaap/_gaap.py:
```diff
@@ -244,12 +251,12 @@ class BaseGaapFluxPlugin(measBase.GenericPlugin):
-
+    def __init__(self, config: BaseGaapFluxConfig, name, schema, logName=None) -> None:
+        """Setup the schema during the initialization of the concrete classes.
+        """
```
We don't usually have docstrings for `__init__` methods.

TallJimbo commented on python/lsst/meas/extensions/gaap/_gaap.py:
```diff
@@ -573,12 +604,93 @@ def fail(self, measRecord, error=None):
+class SingleFrameGaapFluxConfig(BaseGaapFluxConfig,
+                                measBase.SingleFramePluginConfig):
+    """Config for SingleFrameGaapFluxPlugin."""
+    pass
```
I don't think you need `pass` if you have a docstring.

TallJimbo reviewed (approved):
Just a few minor comments here and there.  Overall looks good!

MorganSchmitz reviewed (approved):
And the last commit looks good to me as well!

MorganSchmitz commented on python/lsst/meas/extensions/gaap/_gaap.py:
```diff
@@ -78,46 +81,54 @@ def _isOdd(x: int) -> bool:  # noqa: N805
         dtype=float,
         default=[4.0, 5.0],
-        doc="List of sigmas (in pixels) of circular Gaussian apertures to apply on "
+        doc=("List of sigmas (in pixels) of circular Gaussian apertures to apply on "
             "pre-seeing galaxy images. These should be somewhat larger than the PSF "
             "(determined by ``scalingFactors``) to avoid measurement failures.")
+    )
```
```suggestion
        doc=("List of sigmas (in pixels) of circular Gaussian apertures to apply on "
             "pre-seeing galaxy images. These should be somewhat larger than the PSF "
             "(determined by ``scalingFactors``) to avoid measurement failures.")
    )
```
i.e. there's an extra whitespace at the beggining of ll. 84-85, so the double quotes are vertically aligned with those of l. 83 (instead of being aligned with the opening bracket).

That's only if you decide to keep the extra parenthesis though

arunkannawadi replied:
Ah, sorry, I was somehow blind to this change. I've decided to drop the parenthesis though.

arunkannawadi commented:
I think I've addressed all the comments. One minor additional change is that I added `lint.yaml` file for Github actions, which is mostly a boilerplate file common to all repos. I'll merge this in a couple of hours unless there are objects.

timj commented:
Can you update you GitHub action lint.yaml? It's ever so slightly out of date.

It should look something like:

```yaml
name: lint

on:
  push:
    branches:
      - master
  pull_request:

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2

      - name: Set up Python
        uses: actions/setup-python@v2
        with:
          python-version: 3.8

      - name: Install
        run: pip install -r <(curl https://raw.githubusercontent.com/lsst/linting/master/requirements.txt)

      - name: Run linter
        run: flake8
```
(note the switch to py3.8 and the change to when it runs at the top)

timj commented:
I've enabled standard branch protections on this project but I can't turn on the status check protection until this branch merges (so the action turns up on master)

arunkannawadi commented:
Thanks Tim for enabling the branch protection! I've updated the `lint.yaml` file as well.
